### PR TITLE
feat: refactor and implement warmup

### PR DESCRIPTION
**Pull request by NoobDream2568**
## Ⅰ. Motivation

The initial warmup implementation (PR #335) provided a basic skeleton but lacked policy controls, failure handling, and observability. This PR refines the controller to support retry budgets, global timeouts, and parallelism control.

## Ⅱ. Modifications

**API (`rolebasedgroupwarmup_types.go`):**

- Add `spec.paused` field to suspend/resume the warmup job.
- Add `spec.tolerations` to allow scheduling warmup Pods on tainted nodes (e.g., GPU nodes).
- Expand `WarmupPolicies` with `backoffLimitPerNode`, `maxFailedNodes`, and `globalTimeoutSeconds`.
- Add CEL validation: `maxFailedNodes` requires `backoffLimitPerNode` to be set.
- Add CEL validation: at least one of `imagePreload` or `customizedAction` must be specified in `WarmupActions`.
- Add `Complete` / `Failed` / `VolumeConflict` condition.

**Controller (`rolebasedgroupwarmup_controller.go`):**

- Refactor `reconcileUnfinished` to reduce cyclomatic complexity (extracted `computePermanentlyFailedNodes`, `collectPendingNodes`, `createPodsForNodes`, `requeueForTimeout`).
- Implement `backoffLimitPerNode`: per-node failure counting based on failed Pod labels.
- Implement `maxFailedNodes`: early termination when permanently-failed nodes exceed threshold.
- Implement `globalTimeoutSeconds`: wall-clock timeout from `status.startTime`, with 10-minute max requeue cap.
- Implement `spec.paused`: pause/resume with terminal-state protection (Completed/Failed cannot be overridden).
- Fix label key mismatch: use `constants.GroupNameLabelKey` / `constants.RoleNameLabelKey` instead of deprecated v1alpha1 keys.
- Multi-role merge: deduplicate containers by FNV32 hash (name excluded), detect volume conflicts with `VolumeConflict` condition + warning event.

**Tests (`rolebasedgroupwarmup_controller_test.go`):**

- Add comprehensive unit tests for `buildWarmupPod`, `computePermanentlyFailedNodes`, `collectPendingNodes`, `requeueForTimeout`, `reconcileFinished`, and full `Reconcile` loop scenarios.

**E2E Tests (`test/e2e/testcase/v1alpha2/warmup.go`):**

- Add 6 E2E test cases covering `targetNodes` mode, `targetRoleBasedGroup` multi-role merge, TTL auto-cleanup, global timeout, retry limit + job failure, and no-nodes-matched edge case.
- Add `RoleBasedGroupWarmup` cleanup in framework `AfterEach`.
- Register warmup test cases in `e2e_test.go`.

## Ⅲ. Does this pull request fix one issue?

No.

## Ⅳ. List the added test cases (unit test/integration test) if any, please explain if no tests are needed.

Unit tests added in `rolebasedgroupwarmup_controller_test.go`:

| Test Function                            | Scenarios                                                                       |
|------------------------------------------|---------------------------------------------------------------------------------|
| `TestBuildWarmupPod_ImagePreload`        | Image containers, pullSecrets, tolerations, nodeSelector, labels, restartPolicy |
| `TestBuildWarmupPod_ImageDedup`          | Same image across multiple roles produces one container                         |
| `TestBuildWarmupPod_PullSecretDedup`     | Same secret across roles produces one entry                                     |
| `TestBuildWarmupPod_CustomizedAction`    | Container name override, volume injection                                       |
| `TestBuildWarmupPod_VolumeConflict`      | Same volume name with different specs triggers conflict flag                    |
| `TestBuildWarmupPod_ContainerHashDedup`  | Identical spec with different names deduplicates                                |
| `TestBuildWarmupPod_MixedImageAndCustom` | Mixed mode container indexing                                                   |
| `TestComputePermanentlyFailedNodes_*`    | nil limit, zero limit, exceeded, not exceeded, multi-node                       |
| `TestCollectPendingNodes_*`              | All pending, occupied skipped, permanently failed skipped, mixed, empty         |
| `TestRequeueForTimeout_*`                | No policy, no startTime, short remaining, long remaining (capped)               |
| `TestReconcileFinished_*`                | No TTL, not expired, expired deletes CR                                         |
| `TestReconcile_*`                        | Not found (no-op), paused, resume, completed (no requeue)                       |

E2E tests added in `test/e2e/testcase/v1alpha2/warmup.go`:

| Test Case                                                                         | Scenarios                                                                                                                                                            |
|-----------------------------------------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `complete warmup job with targetNodes mode`                                       | Image preload with `targetNodes`, verify `phase=Completed`, `desired=1`, `succeeded=1`                                                                               |
| `complete warmup job with targetRoleBasedGroup mode and merge multi-role actions` | 2 roles sharing same node, verify multi-role merge produces 2 containers in 1 Pod                                                                                    |
| `auto-delete warmup CR after TTL expires`                                         | `ttlSecondsAfterFinished=10`, verify CR and owned Pods are cascade-deleted                                                                                           |
| `fail warmup job when global timeout is exceeded`                                 | Non-existent image + `globalTimeoutSeconds=30`, verify `phase=Failed`, `reason=GlobalTimeoutExceeded`                                                                |
| `fail warmup job when retry limit is reached`                                     | `backoffLimitPerNode=1` + `maxFailedNodes=0`, container `sleep 3; exit 1`, verify 2 failed Pods → permanent failure → job fails with `reason=MaxFailedNodesExceeded` |
| `complete immediately when no nodes match the target`                             | `nodeSelector` matches no nodes, verify `phase=Completed`, `reason=NoNodesMatched`, `desired=0`                                                                      |

## VI. Special notes for reviews

- `backoffLimitPerNode` counts failed Pods per node. If a cluster's PodGC controller reaps terminated Pods (when `TerminatedPodGCThreshold` is exceeded), the failure count may reset, causing additional retries. This can be improved in the future.
- `globalTimeoutSeconds` does **not** pause when the job is paused — wall-clock time continues to elapse. This is by design.
- Volume conflict across roles on the same node uses first-wins semantics with a `VolumeConflict` warning event and status condition. Pod creation is not blocked.

### CODE REVIEW

File: internal/controller/workloads/rolebasedgroupwarmup_controller.go [hunk 1]
```diff
@@ -92,134 +100,314 @@ func (r *RoleBasedGroupWarmupReconciler) Reconcile(ctx context.Context, req ctrl
...
+func (r *RoleBasedGroupWarmupReconciler) reconcileUnfinished(ctx context.Context, warmup *workloadsv1alpha2.RoleBasedGroupWarmup) (ctrl.Result, error) {
 	logger := log.FromContext(ctx)
-	if err := r.validate(*warmup); err != nil {
-		logger.Error(err, "Validation failed")
-		return err
-	}
 
-	// 1. GetExistingPods
-	// 2. filter pods -> (active pods, succeeded pods, failed pods, )
 	activePods, succeededPods, failedPods, err := r.listWarmupPods(ctx, *warmup)
```
[the remark this answers is not in the document]
Comment (NoobDream2568) [reply internal/controller/workloads/rolebasedgroupwarmup_controller.go:171]: The project uses Go 1.24.1 (go.mod), and since Go 1.22, the loop variable in for _, p := range is per-iteration scoped. Taking &p yields a distinct pointer for each iteration. The classic loop-variable aliasing issue only applies to Go ≤ 1.21.

**Comment by cheyang** [comment]
@gemini-code-assist /review

File: internal/controller/workloads/rolebasedgroupwarmup_controller.go [hunk 2]
```diff
@@ -343,67 +530,79 @@ func (r *RoleBasedGroupWarmupReconciler) listWarmupPods(ctx context.Context, war
...
+			continue
+		}
+		for _, ctr := range action.CustomizedAction.Containers {
+			hashCtr := ctr
+			hashCtr.Name = ""
+			h := fmt.Sprintf("%v", utils.HashContainer(&hashCtr))
+			if ctrHashSet[h] {
+				continue
+			}
+			ctrHashSet[h] = true
```
Comment (cheyang) [inline internal/controller/workloads/rolebasedgroupwarmup_controller.go:590]: Nit: `ctrHashSet` could be `map[uint64]bool` instead of `map[string]bool` to skip the `fmt.Sprintf` allocation per container. Not blocking, just a perf nit for the hot reconciliation path.

**Review by cheyang — APPROVED** [review]
## Review Summary

Solid feature PR that significantly enhances the RoleBasedGroupWarmup controller with parallelism control, backoff/retry logic, global timeout, TTL-based cleanup, and pause/resume support.

**Highlights:**
- Clean controller refactoring with well-separated concerns (createPodsForNodes, computePermanentlyFailedNodes, collectPendingNodes, failWarmupJob, requeueForTimeout)
- Node-level status tracking (vs previous pod-level) is the right abstraction for this use case
- CEL validation on CRD types replaces imperative validation -- good pattern
- Comprehensive test coverage: 819-line unit test suite + 532-line e2e test suite covering happy path, timeout, retry exhaustion, TTL deletion, and empty-match scenarios
- All CI checks pass (lint, unit, envtest, e2e, build)

**Minor observations (non-blocking):**
1. `ctrHashSet` could use `map[uint64]bool` instead of `map[string]bool` to avoid fmt.Sprintf allocations in the reconciliation loop.
2. `NodeNames` target doesn't validate that nodes exist in the cluster -- pods will hang in Pending until timeout. Consider documenting this behavior.

**Note on Gemini's loop variable comment:** The flagged `&p` in `listWarmupPods` is NOT a bug. This project uses Go 1.24.1, and since Go 1.22 each loop iteration creates a new variable -- taking `&p` is safe.

LGTM -- ready to merge.

File: api/workloads/v1alpha2/rolebasedgroupwarmup_types.go [hunk 3]
```diff
@@ -35,51 +32,138 @@ const (
...
+	// RoleBasedGroupWarmup resource and its owned Pods are automatically deleted.
+	// If not set, the resource is NOT auto-deleted.
+	// +optional
+	// +kubebuilder:validation:Minimum=0
+	TTLSecondsAfterFinished *int32 `json:"ttlSecondsAfterFinished,omitempty"`
+}
+
+// RoleBasedGroupWarmupSpec defines the desired state of RoleBasedGroupWarmup.
+// Exactly one of targetNodes or targetRoleBasedGroup must be specified.
+// +kubebuilder:validation:XValidation:rule="has(self.targetNodes) || has(self.targetRoleBasedGroup)",message="either targetNodes or targetRoleBasedGroup must be specified"
```
Comment (cheyang) [inline api/workloads/v1alpha2/rolebasedgroupwarmup_types.go:156]: Since the imperative `validate()` in `reconcileUnfinished` was removed in this PR, these CEL `XValidation` rules are now the **only** layer rejecting invalid specs. CEL validation is GA only on Kubernetes >=1.29 (beta and off-by-default earlier). On clusters below that, an invalid spec reaches the controller — e.g. both `targetNodes` and `targetRoleBasedGroup` nil produces `desired==0` -> job reports `Completed`/`NoNodesMatched`; a `WarmupActions` with neither action makes `buildWarmupPod` emit a zero-container Pod that the API server rejects on Create, retried each reconcile. If rbg's minimum supported k8s version is >=1.29 this is a non-issue; otherwise consider keeping a lightweight guard as a backstop. Just flagging the removed defense layer, non-blocking.

File: internal/controller/workloads/rolebasedgroupwarmup_controller.go [hunk 4]
```diff
@@ -92,134 +100,314 @@ func (r *RoleBasedGroupWarmupReconciler) Reconcile(ctx context.Context, req ctrl
...
-			return err
+	succeededNodeSet := make(map[string]bool)
+	for _, p := range succeededPods {
+		if nodeName, ok := p.Labels[LabelNodeName]; ok {
+			succeededNodeSet[nodeName] = true
 		}
 	}
 
-	return nil
+	failedNodeSet := make(map[string]bool)
```
Comment (cheyang) [inline internal/controller/workloads/rolebasedgroupwarmup_controller.go:381]: Extending Copilot's note on this function: `failedNodeSet` counts **any** node that ever had a failed Pod, so a node that failed once then succeeded ends up in both `succeededNodeSet` and `failedNodeSet`, making `Succeeded + Failed` exceed `Desired`. This is also a different definition of "failed" than `updateStatus`, which counts `permanentlyFailedNodes` (backoff-exhausted). Suggest aligning the two paths — either exclude `succeededNodeSet` from `failedNodeSet` here, or reuse `permanentlyFailedNodes`. Status-counter only, non-blocking.

File: test/e2e/testcase/v1alpha2/warmup.go [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+
+			// Verify warmup Pod was created for the node
+			warmupPodList := &corev1.PodList{}
+			err := f.Client.List(f.Ctx, warmupPodList,
+				client.InNamespace(f.Namespace),
+				client.MatchingLabels{LabelWarmupName: warmup.Name})
+			gomega.Expect(err).ToNot(gomega.HaveOccurred())
+
+			// Should have 1 Pod per unique node (both RBG pods on same node → 1 warmup Pod with 2 containers)
+			gomega.Expect(warmupPodList.Items).To(gomega.HaveLen(1))
```
Comment (cheyang) [inline test/e2e/testcase/v1alpha2/warmup.go:215]: This asserts exactly one warmup Pod, which only holds if both `prefill` and `decode` Pods landed on the same node — but the RBG above sets no affinity/nodeName, and `nodeName` is taken from `podList.Items[0]` only. On a multi-node cluster the two roles can schedule onto different nodes -> 2 desired warmup nodes -> `HaveLen(1)` fails (flaky). Suggest pinning both roles to one node (shared `nodeSelector`/`nodeName` in the pod template, or affinity) so the multi-role-merge intent is deterministic. Same area Copilot flagged.

**Review by cheyang — COMMENTED** [review]
Non-blocking follow-ups complementing the existing bot/human reviews. I verified Gemini's `&p` flag and Copilot's five `*metav1.Time` nil-panic flags are false positives (Go 1.24 per-iteration loop vars; `metav1.Time.IsZero()` is a nil-safe pointer receiver and every `.Time` deref is guarded by `!IsZero()`). The three inline notes below are the items I think are worth acting on — none block merge. Agree with the LGTM.

**Review by cheyang — APPROVED** [review]
/lgtm
/approve
